### ISSUE TITLE
open discussion - say anything here

### ISSUE THREAD

**Issue by zhong-j-yu** [zhong-j-yu/rekex#3]
Leave a comment

**Comment by omegaui** [comment]
@zhong-j-yu 

I am working on an [IDE](https://github.com/omegaui/omegaide)
And i want to write a parser without any extra grammer file.
The IDE currenlty uses a heavy [CodeFramework](https://github.com/omegaui/omegaide/blob/main/src/omega/framework/CodeFramework.java) to parse the source code and i want to write a better (i mean a true parser, since, the CodeFramework is actually not a parser 😶)
While searching the solution I found **rekex**

Sorry, But I am unable to get an idea of what is rekex and how it uses Java 17....?
I mean .. is the grammer defined in source code?
Can you please provide a simple hello world example?

**Comment by zhong-j-yu** [comment]
Hi @omegaui , amazing project!

For a simple example, see https://github.com/zhong-j-yu/rekex/blob/main/rekex-example/src/main/java/org/rekex/exmple/parser/calculator/ExampleParser_Calculator0.java

The entire grammar is defined with datatypes of `Exp0`, `Exp1` etc. Each datatype corresponds to a non-terminal symbol in the grammar.

Then a parser can be constructed by giving the "root datatype", `parser = PegParser.of(Exp0.class)`

You feed an input string to this parser `Exp0 exp0 = parser.matchFull("2+3")`, which will return a parsed structure, in datatype `Exp0`.

Hope that helps. Any more questions are welcome.

**Comment by zhong-j-yu** [comment]
Rekex currently does *not* have a grammar for the Java language. If you want to parse Java files, you probably should look into other solutions. (Not sure if they have support for Java 17 grammar yet)

**Comment by omegaui** [comment]
@zhong-j-yu 

> Hi @omegaui , amazing project!



Thanks.

**Comment by omegaui** [comment]
@zhong-j-yu 

Can you please point out if there is any ready made JavaParser grammer available for parsing java 17 source code?

I checked out antlr's grammers-v4 repo, they had a JavaParser grammer but it's too old, although there is an open issue information them about updating their grammer files but no action has been taken yet!

**Comment by zhong-j-yu** [comment]
Ha, good question. The other Java parser I know of is from Parboiled - https://github.com/sirthias/parboiled/wiki/Java-Parser

But it is very old as well. Java language is so complex it's not an easy task to parse it.

**Comment by omegaui** [comment]
@zhong-j-yu 

Yes, it's a difficult thing.

I got [this](https://github.com/javaparser/javaparser).

It has support for all Java releases till now.

**Comment by mountain** [comment]
@zhong-j-yu Hi, Zhong, I am very glad to see this PEG implementation in Java, and it is very clean. But I also curious about the parsing performance. 

Years ago, I used parboiled to parse log files, and unfortunately found it was slower than regex in C. So at that time,
- we had to switch to python, because regex in python is implemented by C.
- Or, we call unix shells in Hadoop

So, the parsing performance matters in real cases.

**Comment by zhong-j-yu** [comment]
@mountain - thanks, I totally agree with you. In this library, I generate, compile, and load a parser class at runtime; the performance should be comparable to a manually written recursive-decent parser. See https://github.com/zhong-j-yu/rekex/blob/main/rekex-example/src/main/java/org/rekex/exmple/parser/calculator/Generated_ExampleParser_Calculator0.java - there are still little rooms for optimization, but I stopped when it's good enough. At the moment, I'd rather focus on keeping generated Java files readable.

More importantly, Rekex is a way to define grammar in pure Java types and functions. Once a grammar is defined, we could have (but not in the moment) adopt other parsing techniques and algorithms that's best for the use case.

For big log files, I assume each log entry is parsed independently. This library supports parsing a stream of root symbols, see `ParseResult.Partial` in https://github.com/zhong-j-yu/rekex/blob/main/doc/UsersGuide.md#parseresult

**Comment by zhong-j-yu** [comment]
By the way, if a grammar doesn't have recursive rules (I don't think log entries have those), regex is probably the best approach. I'd just personally prefer a more verbose, typed way of defining a regex, see https://github.com/zhong-j-yu/rekex/blob/main/doc/RegExp.md

**Comment by mountain** [comment]
Thanks, Zhong, and yes, I agree with you on the rules of log entries.

As you mentioned 
> generate, compile, and load a parser class at runtime

From the examples, I know we can generate a specific parser at development time, so could you give more docs on the codegen API?

The codegen part is important, especially for cases that license issues matter, when the license of a project is not adapt to Apache2, we can still use rekex by codegen.

**Comment by zhong-j-yu** [comment]
yes you can specify the location of the generated java file with `PegParserBuilder`, see  https://github.com/zhong-j-yu/rekex/blob/main/doc/UsersGuide.md#pegparser

Sorry I don't yet have a ready-to-use maven plugin that generates the java file at build time.

Don't worry about the license, nobody cares. Just treat everything in this repo as in public domain:) If someone really has a problem with Apache2, I'll just relax the license.

**Comment by zhong-j-yu** [comment]
See also method `PegParserBuilder.generateJavaFile()` - https://github.com/zhong-j-yu/rekex/blob/main/rekex-parser/src/main/java/org/rekex/parser/PegParserBuilder.java#L260

**Comment by mountain** [comment]
Thanks, :-)

**Comment by Zireael07** [comment]
The README claims the approach could be ported to "various other languages". Can it be done in, for example, Python or C++?

**Comment by zhong-j-yu** [comment]
@Zireael07 - as long as a language supports algebraic datatype in some sense or in some way.

Even in much older version of Java, it is still possible to do this with annotations or naming conventions to define algebraic datatypes. Of course, that would be unnatural, imposing artifacts in the code for a narrow/special purpose. Once Java 17 supports algebraic datatypes explicitly (albeit with a verbose syntax) this approach becomes easier and more justified -- but still, the end product seems contrived and ugly in places. The basic idea is solid, equating grammar with datatype; the devil is in the details, in how much language features you can exploit for your DSL. In the end, the merit of this kind of library lies in the aesthetics - does it really provide a "better" way to express a grammar.
